File: backend/repositories/repositories/assignment_repository.py
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, and_, func
from sqlalchemy.orm import selectinload
from repositories.models.assignment import Assignment, ProgrammingLanguage
from typing import List, Optional
from datetime import datetime
import sys


class AssignmentRepository:
    """Repository for assignment-related database operations"""

    def __init__(self, db: AsyncSession):
        print(f"[DEBUGGER:AssignmentRepository.__init__:19] db type: {type(db)}", file=sys.stderr, flush=True)
        print(f"[DEBUGGER:AssignmentRepository.__init__:20] db class: {db.__class__.__name__}", file=sys.stderr, flush=True)
        self.db = db

    async def get_assignment_by_id(self, assignment_id: int) -> Optional[Assignment]:
        """
        Get an assignment by ID

        Args:
            assignment_id: ID of the assignment

        Returns:
            Assignment object or None if not found
        """
        result = await self.db.execute(
            select(Assignment).where(Assignment.id == assignment_id)
        )
        return result.scalar_one_or_none()

    async def get_assignments_by_class(
        self,
        class_id: int,
        active_only: bool = True
    ) -> List[Assignment]:
        """
        Get all assignments for a class

        Args:
            class_id: ID of the class
            active_only: If True, only return active assignments

        Returns:
            List of Assignment objects
        """
        query = select(Assignment).where(Assignment.class_id == class_id)

        if active_only:
            query = query.where(Assignment.is_active == True)

        query = query.order_by(Assignment.deadline.asc())
        result = await self.db.execute(query)
        return list(result.scalars().all())

    async def get_pending_assignments_by_teacher(
        self,
        teacher_id: int,
        limit: Optional[int] = None
    ) -> List[Assignment]:
        """
        Get pending assignments for all classes taught by a teacher
        Returns assignments that haven't passed their deadline yet

        Args:
            teacher_id: ID of the teacher
            limit: Optional limit on number of assignments to return

        Returns:
            List of Assignment objects
        """
        from repositories.models.class_model import Class

        query = (
            select(Assignment)
            .options(selectinload(Assignment.class_obj))
            .join(Class, Assignment.class_id == Class.id)
            .where(and_(
                Class.teacher_id == teacher_id,
                Assignment.is_active == True,
                Assignment.deadline >= func.now()
            ))
            .order_by(Assignment.deadline.asc())
        )

        if limit:
            query = query.limit(limit)

        result = await self.db.execute(query)
        return list(result.scalars().all())

    async def get_assignments_needing_review_by_teacher(
        self,
        teacher_id: int,
        limit: Optional[int] = None
    ) -> List[Assignment]:
        """
        Get assignments that need review for all classes taught by a teacher
        This is a placeholder - you'll need to implement this when you have
        a submissions table to check which assignments have ungraded submissions

        Args:
            teacher_id: ID of the teacher
            limit: Optional limit on number of assignments to return

        Returns:
            List of Assignment objects
        """
        print(f"[DEBUGGER:AssignmentRepository.get_assignments_needing_review_by_teacher:115] teacher_id={teacher_id}, limit={limit}", file=sys.stderr, flush=True)
        print(f"[DEBUGGER:AssignmentRepository.get_assignments_needing_review_by_teacher:116] self.db type: {type(self.db)}", file=sys.stderr, flush=True)
        from repositories.models.class_model import Class

        # TODO: Join with submissions table to filter assignments with pending reviews
        # For now, return all active assignments
        print(f"[DEBUGGER:AssignmentRepository.get_assignments_needing_review_by_teacher:121] Building query...", file=sys.stderr, flush=True)
        query = (
            select(Assignment)
            .options(selectinload(Assignment.class_obj))
            .join(Class, Assignment.class_id == Class.id)
            .where(and_(
                Class.teacher_id == teacher_id,
                Assignment.is_active == True
            ))
            .order_by(Assignment.deadline.asc())
        )

        if limit:
            query = query.limit(limit)

        print(f"[DEBUGGER:AssignmentRepository.get_assignments_needing_review_by_teacher:136] About to execute query", file=sys.stderr, flush=True)
        result = await self.db.execute(query)
        print(f"[DEBUGGER:AssignmentRepository.get_assignments_needing_review_by_teacher:138] Query executed successfully", file=sys.stderr, flush=True)
        return list(result.scalars().all())

    async def create_assignment(
        self,
        class_id: int,
        assignment_name: str,
        description: str,
        programming_language: ProgrammingLanguage,
        deadline: datetime,
        allow_resubmission: bool = True
    ) -> Assignment:
        """
        Create a new assignment

        Args:
            class_id: ID of the class
            assignment_name: Name of the assignment
            description: Assignment description
            programming_language: Programming language for the assignment
            deadline: Deadline for submission
            allow_resubmission: Whether to allow resubmissions

        Returns:
            Created Assignment object
        """
        new_assignment = Assignment(
            class_id=class_id,
            assignment_name=assignment_name,
            description=description,
            programming_language=programming_language,
            deadline=deadline,
            allow_resubmission=allow_resubmission,
            is_active=True
        )

        self.db.add(new_assignment)
        await self.db.commit()
        await self.db.refresh(new_assignment)

        return new_assignment

    async def update_assignment(
        self,
        assignment_id: int,
        **kwargs
    ) -> Optional[Assignment]:
        """
        Update an assignment

        Args:
            assignment_id: ID of the assignment to update
            **kwargs: Fields to update

        Returns:
            Updated Assignment object or None if not found
        """
        assignment = await self.get_assignment_by_id(assignment_id)

        if not assignment:
            return None

        for key, value in kwargs.items():
            if hasattr(assignment, key):
                setattr(assignment, key, value)

        await self.db.commit()
        await self.db.refresh(assignment)

        return assignment

    async def delete_assignment(self, assignment_id: int) -> bool:
        """
        Delete an assignment (hard delete)

        Args:
            assignment_id: ID of the assignment to delete

        Returns:
            True if deleted, False if not found
        """
        assignment = await self.get_assignment_by_id(assignment_id)

        if not assignment:
            return False

        await self.db.delete(assignment)
        await self.db.commit()

        return True

    async def get_pending_assignments_for_student(
        self,
        student_id: int,
        limit: Optional[int] = None
    ) -> List[Assignment]:
        """
        Get pending assignments for all classes a student is enrolled in
        Returns assignments that haven't passed their deadline yet

        Args:
            student_id: ID of the student
            limit: Optional limit on number of assignments to return

        Returns:
            List of Assignment objects
        """
        from repositories.models.class_model import Class
        from repositories.models.enrollment import Enrollment

        query = (
            select(Assignment)
            .options(selectinload(Assignment.class_obj))
            .join(Class, Assignment.class_id == Class.id)
            .join(Enrollment, Class.id == Enrollment.class_id)
            .where(and_(
                Enrollment.student_id == student_id,
                Assignment.is_active == True,
                Class.is_active == True,
                Assignment.deadline >= func.now()
            ))
            .order_by(Assignment.deadline.asc())
        )

        if limit:
            query = query.limit(limit)

        result = await self.db.execute(query)
        return list(result.scalars().all())

File: backend-ts/src/repositories/assignment.repository.ts
import { db } from "@/shared/database.js";
import { eq, and, desc, inArray, sql, or, gt, asc } from "drizzle-orm";
import {
  assignments,
  classes,
  submissions,
  enrollments,
  type Assignment,
  type NewAssignment,
  type LatePenaltyConfig,
} from "@/models/index.js";
import { BaseRepository } from "@/repositories/base.repository.js";
import { injectable } from "tsyringe";
/** Programming language type */
export type ProgrammingLanguage = "python" | "java" | "c";

/**
 * Repository for assignment-related database operations.
 */
@injectable()
export class AssignmentRepository extends BaseRepository<
  typeof assignments,
  Assignment,
  NewAssignment
> {
  constructor() {
    super(assignments);
  }

  /** Get an assignment by ID */
  async getAssignmentById(
    assignmentId: number
  ): Promise<Assignment | undefined> {
    return await this.findById(assignmentId);
  }

  /** Get all assignments for a class */
  async getAssignmentsByClassId(
    classId: number,
    activeOnly: boolean = true
  ): Promise<Assignment[]> {
    if (activeOnly) {
      return await this.db
        .select()
        .from(assignments)
        .where(
          and(eq(assignments.classId, classId), eq(assignments.isActive, true))
        )
        .orderBy(desc(assignments.deadline));
    }

    return await this.db
      .select()
      .from(assignments)
      .where(eq(assignments.classId, classId))
      .orderBy(desc(assignments.deadline));
  }

  /** Get assignments for multiple classes */
  async getAssignmentsByClassIds(
    classIds: number[],
    activeOnly: boolean = true
  ): Promise<Assignment[]> {
    if (classIds.length === 0) {
      return [];
    }

    if (activeOnly) {
      return await this.db
        .select()
        .from(assignments)
        .where(
          and(
            inArray(assignments.classId, classIds),
            eq(assignments.isActive, true)
          )
        )
        .orderBy(desc(assignments.deadline));
    }

    return await this.db
      .select()
      .from(assignments)
      .where(inArray(assignments.classId, classIds))
      .orderBy(desc(assignments.deadline));
  }

  /**
   * Get pending tasks for a teacher.
   * Combines logic to check for upcoming deadlines or ungraded submissions.
   * Optimized to avoid N+1 query problem.
   */
  async getPendingTasksForTeacher(
    teacherId: number,
    limit: number = 10
  ): Promise<
    Array<{
      id: number;
      assignmentName: string;
      deadline: Date | null;
      classId: number;
      className: string;
      studentCount: number;
      submissionCount: number;
    }>
  > {
    const now = new Date();

    // Subquery for student counts per class
    const classStudentCounts = this.db
      .select({
        classId: enrollments.classId,
        count: sql<number>`count(*)`.as("studentCount"),
      })
      .from(enrollments)
      .groupBy(enrollments.classId)
      .as("csc");

    // Subquery for latest submission counts per assignment
    const assignmentSubmissionCounts = this.db
      .select({
        assignmentId: submissions.assignmentId,
        count: sql<number>`count(*)`.as("submissionCount"),
      })
      .from(submissions)
      .where(eq(submissions.isLatest, true))
      .groupBy(submissions.assignmentId)
      .as("asc");

    const results = await this.db
      .select({
        id: assignments.id,
        assignmentName: assignments.assignmentName,
        deadline: assignments.deadline,
        classId: classes.id,
        className: classes.className,
        studentCount: sql<number>`COALESCE(${classStudentCounts.count}, 0)`,
        submissionCount: sql<number>`COALESCE(${assignmentSubmissionCounts.count}, 0)`,
      })
      .from(assignments)
      .innerJoin(classes, eq(assignments.classId, classes.id))
      .leftJoin(classStudentCounts, eq(classes.id, classStudentCounts.classId))
      .leftJoin(
        assignmentSubmissionCounts,
        eq(assignments.id, assignmentSubmissionCounts.assignmentId)
      )
      .where(
        and(
          eq(classes.teacherId, teacherId),
          eq(classes.isActive, true),
          eq(assignments.isActive, true),
          or(
            gt(assignments.deadline, now),
            gt(sql`COALESCE(${assignmentSubmissionCounts.count}, 0)`, 0)
          )
        )
      )
      .orderBy(asc(assignments.deadline))
      .limit(limit);

    return results.map((r) => ({
      ...r,
      studentCount: Number(r.studentCount),
      submissionCount: Number(r.submissionCount),
      deadline: r.deadline ? new Date(r.deadline) : null,
    }));
  }

  /** Create a new assignment */
  async createAssignment(data: {
    classId: number;
    assignmentName: string;
    description: string;
    programmingLanguage: ProgrammingLanguage;
    deadline: Date;
    allowResubmission?: boolean;
    maxAttempts?: number | null;
    templateCode?: string | null;
    totalScore?: number;
    scheduledDate?: Date | null;
  }): Promise<Assignment> {
    const results = await this.db
      .insert(assignments)
      .values({
        classId: data.classId,
        assignmentName: data.assignmentName,
        description: data.description,
        programmingLanguage: data.programmingLanguage,
        deadline: data.deadline,
        allowResubmission: data.allowResubmission ?? true,
        maxAttempts: data.maxAttempts ?? null,
        templateCode: data.templateCode ?? null,
        totalScore: data.totalScore ?? 100,
        scheduledDate: data.scheduledDate ?? null,
        isActive: true,
      })
      .returning();

    return results[0];
  }

  /** Update an assignment */
  async updateAssignment(
    assignmentId: number,
    data: Partial<
      Pick<
        NewAssignment,
        | "assignmentName"
        | "description"
        | "programmingLanguage"
        | "deadline"
        | "allowResubmission"
        | "maxAttempts"
        | "isActive"
        | "templateCode"
        | "totalScore"
        | "scheduledDate"
      >
    >
  ): Promise<Assignment | undefined> {
    const updateData = Object.fromEntries(
      Object.entries(data).filter(([_, v]) => v !== undefined)
    );

    if (Object.keys(updateData).length === 0) {
      return await this.getAssignmentById(assignmentId);
    }

    return await this.update(assignmentId, updateData);
  }

  /** Delete an assignment (hard delete) */
  async deleteAssignment(assignmentId: number): Promise<boolean> {
    return await this.delete(assignmentId);
  }

  /** Deactivate an assignment (soft delete) */
  async deactivateAssignment(
    assignmentId: number
  ): Promise<Assignment | undefined> {
    return await this.update(assignmentId, { isActive: false });
  }

  /**
   * Set late penalty configuration for an assignment.
   */
  async setLatePenaltyConfig(
    assignmentId: number,
    enabled: boolean,
    config: LatePenaltyConfig | null
  ): Promise<void> {
    await this.db
      .update(assignments)
      .set({
        latePenaltyEnabled: enabled,
        latePenaltyConfig: config,
      })
      .where(eq(assignments.id, assignmentId));
  }

  /**
   * Get late penalty configuration for an assignment.
   * Returns null if no config is set.
   */
  async getLatePenaltyConfig(
    assignmentId: number
  ): Promise<{ enabled: boolean; config: LatePenaltyConfig | null } | null> {
    const result = await this.db
      .select({
        latePenaltyEnabled: assignments.latePenaltyEnabled,
        latePenaltyConfig: assignments.latePenaltyConfig,
      })
      .from(assignments)
      .where(eq(assignments.id, assignmentId))
      .limit(1);

    if (!result[0]) return null;

    return {
      enabled: result[0].latePenaltyEnabled,
      config: result[0].latePenaltyConfig,
    };
  }
}
